File: android/app/src/main/java/eu/id3/bioseal/sample/Sample.kt
package eu.id3.bioseal.sample

import android.content.Context
import eu.id3.bioseal.BiometricDataType
import eu.id3.bioseal.Bioseal
import eu.id3.bioseal.BiosealError
import eu.id3.bioseal.BiosealException
import eu.id3.bioseal.FieldExtensionType
import eu.id3.bioseal.FieldType
import eu.id3.bioseal.ResourceCallbackArgs
import java.io.File
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date

class Sample(val context: Context) {
    val bioseal: Bioseal by lazy { Bioseal() }
    val localCache = true
    val dateFormat: DateFormat = SimpleDateFormat("EEEE, MMMM d, yyyy")
    val timeFormat: DateFormat = SimpleDateFormat("h:mm a")
    val localCacheDir = "/storage/emulated/0/Download/cache_bioseal"

    fun run() {
        println("-------------------")
        println("id3.BioSeal.Samples")
        println("-------------------")

        // The bioseal instance must first be initialized
        bioseal.setExternalResourceCallback(::getExternalResourceWithCache)

        // This basic sample shows how to read BioSeal biographics only contents
        displayBioSealInfo("ExBioSealBiographics.dat")

        // This sample shows how to read BioSeal face image and template contents
        displayBioSealInfo("ExBioSealFace.dat")

        // This accreditation sample shows how to read BioSeal face template contents
        displayBioSealInfo("ExBioSealAccreditation.dat")
    }

    private fun displayBioSealInfo(path: String) {
        println("Decoding BioSeal file $path")
        val dataBioseal = context.assets.open(path).readBytes()
        try {
            bioseal.verifyFromBuffer(dataBioseal)
        } catch (ex: BiosealException) {
            println(ex.message)
            throw ex
        }

        println("  BioSeal format : ${bioseal.format}")
        // display manifest information
        println("   Use case: ${bioseal.manifestId.toString(16)}")
        val supportedLanguages = bioseal.supportedLanguages
        for (i in 0 until supportedLanguages.count) {
            println("   Document name: '${bioseal.getDocumentName(supportedLanguages[i])}'")
        }

        println( "   Payload:")
        val payload = getPayload()
        for ((key, value) in payload.entries) {
            println("      $key: $value")
        }

        // display face template if existing
        val hasFaceTemplate = bioseal.containsFaceTemplates
        println("   Face template: $hasFaceTemplate")
        if (hasFaceTemplate) {
            val fieldFaceList =
                bioseal.findBiometrics(BiometricDataType.FACIAL_FEATURES, null)
            if (fieldFaceList.count > 0) {
                val fieldFace = fieldFaceList.get(0)
                println("   Face template saved in data folder")
                File(
                    localCacheDir,
                    path.replace(".dat", ".template")
                ).writeBytes(fieldFace.valueAsBinary)
            }
        }

        // display and save face image if existing
        val hasFaceImage = bioseal.containsPortraits
        println("   Face image: $hasFaceImage")
        if (hasFaceImage) {
            val fieldFaceList =
                bioseal.findFieldsByExtension(FieldExtensionType.PORTRAIT)
            if (fieldFaceList.count > 0) {
                val fieldFace = fieldFaceList.get(0)
                println("   Face image saved in data folder")
                File(
                    localCacheDir,
                    path.replace(".dat", ".webp")
                ).writeBytes(fieldFace.valueAsBinary)
            }
        }

        // fetch and save presentation view
        println("   Presentation view supported languages:")
        for (i in 0 until bioseal.supportedLanguages.count) {
            println("      ${bioseal.supportedLanguages[i]}")
        }

        // Get default language
        bioseal.buildHtmlView(null, true)
        File(localCacheDir, path.replace(".dat", ".html")).writeText(bioseal.htmlView)
        println("   Presentation view saved in data folder")

        // build and save JSON file
        val jsonPayload = bioseal.buildVdsAsJson("  ")
        File(localCacheDir, path.replace(".dat", ".json")).writeText(jsonPayload)
        println("   JSON representation saved in data folder")

        // display signature information
        println("   Signature:")
        println("      Signature verified status: ${bioseal.verificationResult.vdsSignatureVerified}")
        println("      Certification chain verified status: ${bioseal.verificationResult.certificationChainVerified}")
        println("      Certificate usage authorized status: ${bioseal.verificationResult.signingCertificateUsageAuthorized}")

        // display governance information
        println("   Governance:")
        println("      LoTL: ${bioseal.lotlUrl}")
        println("      TSL: ${bioseal.tslUrl}")
        println("      Manifest: ${bioseal.manifestUrl}")
        println("      LoTL valid status: ${bioseal.verificationResult.lotlGovernanceValid}")
        println("      TSL valid status: ${bioseal.verificationResult.tslGovernanceValid}")
        println("      Manifest valid status: ${bioseal.verificationResult.manifestGovernanceValid}")
        println("      Authority verified status: ${bioseal.verificationResult.caCertificateVerified}")

        // display certificate information
        println("   Certificate:")
        println("      Authority AC: ${bioseal.certificateAuthorityReference}")
        println("      Authority ID: ${bioseal.certificateIdentifier}")
        println("      Issuer: ${bioseal.certificateInformation.issuerCommonName}")
        println("      Subject: ${bioseal.certificateInformation.subjectCommonName}")
        println("      Organization: ${bioseal.certificateInformation.subjectOrganization}")
        println("      Organization unit: ${bioseal.certificateInformation.subjectOrganizationalUnit}")
        println("      Date of creation: ${bioseal.certificateInformation.notBefore.toString()}")
        println("      Date of expiration: ${bioseal.certificateInformation.notAfter.toString()}")
    }

    /// Gets the dictionary of biographics data from the BioSeal instance.
    private fun getPayload(): Map<String, Any> {
        val payload = mutableMapOf<String, Any>()
        // Scan payload
        var data = "?"
        for (i in 0 until bioseal.payload.keys.count) {
            val field = bioseal.payload.get(bioseal.payload.keys[i])
            data = when (field.fieldType) {
                FieldType.INTEGER -> field.valueAsInteger.toString()
                FieldType.BOOLEAN -> field.valueAsBoolean.toString()
                FieldType.FLOAT -> field.valueAsFloat.toString()
                FieldType.STRING -> field.valueAsString
                FieldType.BINARY -> String(field.valueAsBinary).replace("-", "")
                FieldType.DATE -> {
                    val date = field.valueAsDate
                    val dateTime = Date.UTC(date.year, date.month, date.day, 0, 0, 0)
                    dateFormat.format(dateTime)
                }

                FieldType.TIME -> {
                    val time = field.valueAsTime
                    val now = Calendar.getInstance().time
                    val dateTime =
                        Date.UTC(now.year, now.month, now.day, time.hour, time.minute, time.second)
                    timeFormat.format(dateTime)
                }

                FieldType.TIMESTAMP -> {
                    val timestamp = field.valueAsDateTime
                    val dateTime = Date.UTC(
                        timestamp.year,
                        timestamp.month,
                        timestamp.day,
                        timestamp.hour,
                        timestamp.minute,
                        timestamp.second
                    )
                    dateTime.toString()
                }

                else -> "?"
            }
            payload[field.name] = data
        }
        return payload
    }

    /// Implements the GetExternalResource with cache callback.
    private fun getExternalResourceWithCache(obj: Any, args: ResourceCallbackArgs): Int {
        var err = 0
        try {
            if (!localCache) {
                args.download()
            } else {
                println("URI = ${args.uri}")
                // Cache disque
                if (!File(localCacheDir).exists()) {
                    File(localCacheDir).mkdir()
                }

                val cacheFile = File(localCacheDir, args.resourceName)
                if (!cacheFile.exists() || args.requiresUpdate) {
                    try {
                        args.download()
                        cacheFile.createNewFile()
                        cacheFile.writeBytes(args.outputData)
                    } catch (ex: BiosealException) {
                        err = BiosealError.RESOURCE_NOT_FOUND.value
                    }
                } else {
                    args.outputData = cacheFile.readBytes()
                }
            }
        } catch (ex: BiosealException) {
            println(ex.message)
            err = BiosealError.EXCEPTION_IN_CALLBACK.value
        }
        return err
    }

}

File: ios/Bioseal Sample/Bioseal Sample/Sample.swift
import Foundation
import UIKit
import id3Bioseal

class Sample {
    let bioseal: Bioseal = try! Bioseal()
    let localCache = true
    let dateFormat: DateFormatter = DateFormatter()
    let timeFormat: DateFormatter = DateFormatter()
    let downloadsDirectory: URL
    let localCacheDir: URL

    init() {
        self.dateFormat.dateFormat = "EEEE, MMMM d, yyyy"
        self.timeFormat.dateFormat = "h:mm a"
        downloadsDirectory = FileManager.default.urls(for: .downloadsDirectory, in: .userDomainMask).first!
        localCacheDir = downloadsDirectory.appendingPathComponent("cache_bioseal")
    }

    func run() {
        print("-------------------")
        print("id3.BioSeal.Samples")
        print("-------------------")

        // This basic sample shows how to read BioSeal biographics only contents
        try! displayBioSealInfo(name: "ExBioSealBiographics")

        // This sample shows how to read BioSeal face image and template contents
        try! displayBioSealInfo(name: "ExBioSealFace")

        // This accreditation sample shows how to read BioSeal face template contents
        try! displayBioSealInfo(name: "ExBioSealAccreditation")
    }

    private func displayBioSealInfo(name: String) throws {
        print("Decoding BioSeal file \(name).dat")
        let dataBioseal = [UInt8](NSDataAsset(name: name)!.data)

        do {
            try bioseal.verifyFromBuffer(data: dataBioseal)
            try bioseal.verify()
        } catch let ex as BiosealException {
            print(ex.getMessage())
        }

        print("  BioSeal format : \(try bioseal.getFormat())")
        // display manifest information
        print("   Use case: \(String(format: "%X", try bioseal.getManifestId()))")
        let supportedLanguages = try bioseal.getSupportedLanguages()
        let supportedLanguagesCount = try supportedLanguages.getCount()
        for i in 0..<supportedLanguagesCount {
            print("   Document name: '\(try bioseal.getDocumentName(language: try supportedLanguages.get(index: i)))'")
        }

        print("   Payload:")
        let payload = try getPayload()
        for (key, value) in payload {
            print("      \(key): \(value)")
        }

        // display face template if existing
        let hasFaceTemplate = try bioseal.getContainsFaceTemplates()
        print("   Face template: \(hasFaceTemplate)")
        if hasFaceTemplate {
            let fieldFaceList = try bioseal.findBiometrics(biometricDataType: BiometricDataType.facialFeatures, biometricFormat: "")
            if(try fieldFaceList.getCount() > 0) {
                let fieldFace = try fieldFaceList.get(index: 0)
                print("   Face template saved in data folder")
                let templateFilePath = localCacheDir.appendingPathComponent(name + ".template").path
                FileManager.default.createFile(atPath: templateFilePath, contents: Data(try fieldFace.getValueAsBinary()), attributes: nil)
            }
        }

        // display and save face image if existing
        let hasFaceImage = try bioseal.getContainsPortraits()
        print("   Face image: \(hasFaceImage)")
        if hasFaceImage {
            let fieldFaceList = try bioseal.findBiometrics(biometricDataType: BiometricDataType.facialFeatures, biometricFormat: "")
            if (try fieldFaceList.getCount() > 0) {
                let fieldFace = try fieldFaceList.get(index: 0)
                print("   Face image saved in data folder")
                let imageFilePath = localCacheDir.appendingPathComponent(name + ".webp").path
                FileManager.default.createFile(atPath: imageFilePath, contents: Data(try fieldFace.getValueAsBinary()), attributes: nil)
            }
        }

        // fetch and save presentation view
        print("   Presentation view supported languages:")
        let supportedHtmlViewLanguages = try bioseal.getSupportedHtmlViewLanguages()
        let supportedHtmlViewLanguagesCount = try supportedHtmlViewLanguages.getCount()
        for i in 0..<supportedHtmlViewLanguagesCount {
            print("      \(try supportedHtmlViewLanguages.get(index: i))")
        }

        // Get default language
        try bioseal.buildHtmlView(language: "en", userAuthenticated: true)
        let htmlFilePath = localCacheDir.appendingPathComponent(name + ".html").path
        try? bioseal.getHtmlView().write(toFile: htmlFilePath, atomically: true, encoding: .utf8)

        print("   Presentation view saved in data folder")

        // build and save JSON file
        let jsonPayload = try bioseal.buildVdsAsJson(indentation: "  ")
        let jsonFilePath = localCacheDir.appendingPathComponent(name +  ".json").path
        try? jsonPayload.write(toFile: jsonFilePath, atomically: true, encoding: .utf8)
        print("   JSON representation saved in data folder")

        // display signature information
        print("   Signature:")
        print("      Signature verified status: \(try bioseal.getVerificationResult().VdsSignatureVerified)")
        print("      Certification chain verified status: \(try bioseal.getVerificationResult().CertificationChainVerified)")
        print("      Certificate usage authorized status: \(try bioseal.getVerificationResult().SigningCertificateUsageAuthorized)")

        // display governance information
        print("   Governance:")
        print("      LoTL: \(try bioseal.getLotlUrl())")
        print("      TSL: \(try bioseal.getTslUrl())")
        print("      Manifest: \(try bioseal.getManifestUrl())")
        print("      LoTL valid status: \(try bioseal.getVerificationResult().LotlGovernanceValid)")
        print("      TSL valid status: \(try bioseal.getVerificationResult().TslGovernanceValid)")
        print("      Manifest valid status: \(try bioseal.getVerificationResult().ManifestGovernanceValid)")
        print("      Authority verified status: \(try bioseal.getVerificationResult().CaCertificateVerified)")

        // display certificate information
        print("   Certificate:")
        print("      Authority reference: \(try bioseal.getCertificateAuthorityReference())")
        print("      Issuer: \(try bioseal.getCertificateInformation().getIssuerCommonName())")
        print("      Subject: \(try bioseal.getCertificateInformation().getSubjectCommonName())")
        print("      Organization: \(try bioseal.getCertificateInformation().getSubjectOrganization())")
        print("      Organization unit: \(try bioseal.getCertificateInformation().getSubjectOrganizationalUnit())")
        print("      Date of creation: \(try bioseal.getCertificateInformation().getNotBefore())")
        print("      Date of expiration: \(try bioseal.getCertificateInformation().getNotAfter())")
    }

    /// Gets the dictionary of biographics data from the BioSeal instance.
    private func getPayload() throws -> [String: Any] {
        var payload: [String: Any] = [:]
        // Scan payload
        var data: Any = "?"
        let payloadKeys = try bioseal.getPayload().getKeys()
        let payloadKeysCount = try payloadKeys.getCount()
        for i in 0..<payloadKeysCount {
            let field = try bioseal.getPayload().get(key: try payloadKeys.get(index: i))
            switch try field.getFieldType() {
            case .integer:
                data = try field.getValueAsInteger()
            case .boolean:
                data = try field.getValueAsBoolean()
            case .float:
                data = try field.getValueAsFloat()
            case .string:
                data = try field.getValueAsString()
            case .binary:
                data = String(bytes: try field.getValueAsBinary(), encoding: .utf8)?.replacingOccurrences(of: "-", with: "")
            case .date:
                let date = try field.getValueAsDate()
                let dateTime = Calendar.current.date(from: DateComponents(year: Int(try date.getYear()), month: Int(try date.getMonth()), day: Int(try date.getDay())))
                data = dateFormat.string(from: dateTime!)
            case .time:
                let time = try field.getValueAsTime()
                let now = Calendar.current.dateComponents(in: .current, from: Date())
                let dateTime = Calendar.current.date(from: DateComponents(year: now.year, month: now.month, day: now.day, hour: Int(try time.getHour()), minute: Int(try time.getMinute()), second: Int(try time.getSecond())))
                data = timeFormat.string(from: dateTime!)
            case .timestamp:
                let timestamp = try field.getValueAsDateTime()
                let dateTime = Calendar.current.date(from: DateComponents(year: Int(try timestamp.getYear()), month: Int(try timestamp.getMonth()), day: Int(try timestamp.getDay()), hour: Int(try timestamp.getHour()), minute: Int(try timestamp.getMinute()), second: Int(try timestamp.getSecond())))
                data = dateTime?.description ?? "?"
            default:
                data = "?"
            }
            payload[try field.getName()] = data
        }
        return payload
    }
}
